### DISCUSSION TITLE
Support CSS View Transitions in `wire:navigate`

### DISCUSSION THREAD

**Discussion by jacksleight** [livewire/livewire#6329] [3. Feature ideas]
It would be awesome if you could use [CSS View Transitions](https://developer.mozilla.org/en-US/docs/Web/API/View_Transitions_API) with `wire:navigate`.

Could be a modifier on the attribute, but it might also make sense as a simple global option that just enables transitions for any navigate calls.

**Comment by iksaku** [comment]
Looking at the current implementation via [Alpine plugin `x-navigate`](https://github.com/livewire/livewire/blob/750eb122889269b558318bcb269ecf42fe27d48b/js/plugins/navigate/index.js) we can add a function to enable the View Transitions API:

```js
let enablePersist = true
let showProgressBar = true
let restoreScroll = true
let autofocus = false
let useTransitionApi = false

export default function (Alpine) {
    Alpine.navigate = (url) => {
        navigateTo(
            createUrlObjectFromString(url)
        )
    }

    Alpine.navigate.disableProgressBar = () => {
        showProgressBar = false
    }

    Alpine.navigate.enableTransitionApi = () => {
        useTransitionApi = true
    }

    // ...
}
```

In theory, now we can wrap calls to `swapCurrentPageWithNewHtml()` [when navigating forward](https://github.com/livewire/livewire/blob/750eb122889269b558318bcb269ecf42fe27d48b/js/plugins/navigate/index.js#L68-L82) and [when navigating backwards](https://github.com/livewire/livewire/blob/750eb122889269b558318bcb269ecf42fe27d48b/js/plugins/navigate/index.js#L96-L108) to have a [basic view transition](https://developer.mozilla.org/en-US/docs/Web/API/View_Transitions_API#creating_a_basic_view_transition) implementation:

```js
let theFunBegin = () => swapCurrentPageWithNewHtml(html, (andAfterAllThis) => { /* ... */ })

if (!useTransitionApi || ! document.startViewTransition) {
    theFunBegin()
    return
}

document.startViewTransition(theFunBegin)
```

Another option could be to change the signature of [`swapCurrentPageWithNewHtml()`](https://github.com/livewire/livewire/blob/750eb122889269b558318bcb269ecf42fe27d48b/js/plugins/navigate/page.js#L7-L34):

```js
export function swapCurrentPageWithNewHtml(html, useTransitionApi, andThen) {
    let theFunBegin = () => { /* Current function implementation... Just wrapped */ }

    if (!useTransitionApi || ! document.startViewTransition) {
        theFunBegin()
        return
    }

    document.startViewTransition(theFunBegin)
}
```

<hr/>

As an extra, I think Astro has a good implementation of the View Transition API using their own view components as opt-in option for specific pages:
* [Astro's Docs on View Transitions](https://docs.astro.build/en/guides/view-transitions/).
* [Astro's `<ViewTransitions />` component](https://github.com/withastro/astro/blob/3674584e02b161a698b429ceb66723918fdc56ac/packages/astro/components/ViewTransitions.astro)

**Reply by jacksleight** [reply]
Sounds perfect! If I can help with a PR or testing or anything just say. 🙂

**Reply by iksaku** [reply]
I will look more deeply into how the implementation is done by other frameworks/libraries before starting a PR.

I just came across the new SvelteKit release and the implementation they've gone with looks like could work with Livewire pretty similarly. Although it is not a _magic integration_ like Astro's, the _manual_ way they've gone with is still pretty much plug-n-play and also allows more flexibility for the developer.

Here's the Svelte blog post detailing their new API for implementing View Transitions: https://svelte.dev/blog/view-transitions.

**Reply by iksaku** [reply]
Made a quick proof of concept adapting the SvelteKit implementation into the Alpine navigate plugin. Works pretty good!

https://github.com/livewire/livewire/assets/4632429/80229975-e4eb-43bd-9c6c-3741512fb839

As this takes a more manual approach, the developer would need to hook into a `beforeDomUpdate()` callback that will enable View Transition lifecycle:
```js
// Both `alpine:navigating` and `livewire:navigating` expose the same hook.
document.addEventListener('livewire:navigating', (ev) => {
    if (!document.startViewTransition) return

    ev.detail.beforeDomUpdate((navigation) => {
        return new Promise((resolve) => {
            document.startViewTransition(async () => {
                resolve()
                await navigation.complete
            })
        })
    })
})
```

You can then link your elements for the transition with CSS:
```css
#title {
    view-transition-name: title;
}

#publish {
    view-transition-name: publish;
}
```

The current POC does a per-event hooking, meaning that there should be leaking memory as the scope is cleaned up once the navigation finishes executing, but I'm still not fan of the `beforeDomUpdate()` function requiring to hook into the `alpine:navigating|livewire:navigating` event... Probably something like `Alpine.onNavigate()` would be better (and the name being the same as SvelteKit... I'm bad at naming things), then my only concern would be the probability of others ignoring the `data-navigate-track` attribute in their scripts and infinitely hooking the same code forever.

Any thoughts?

**Reply by iksaku** [reply]
Pushed the implementation code to [`iksaku/livewire/on-navigate`](https://github.com/iksaku/livewire/tree/feature/on-navigate), also, just pushed a change to hook into the late navigation event using `Alpine.navigate.onNavigate()`.

Any function that should hook into this function should do so after Alpine initialization, either at `alpine:init` or `alpine:navigated`:
```js
document.addEventListener('alpine:init', () = {
    Alpine.navigate.onNavigate((navigation) => {
        return new Promise((resolve) => {
            document.startViewTransition(async () => {
                resolve()
                await navigation.complete
            })
        })
    })
})
```

---

Update: Changed branch URL to a clean implementation one.

**Reply by jacksleight** [reply]
Nice! I think that's a great solution, and like you say doing it with a new `onNavigate` event like Svelte allows for a lot more flexibility than just a fixed `startViewTransition` call.

**Reply by iksaku** [reply]
Just uploaded the PR for this feature. Check it out: #6893

**Reply by iksaku** [reply]
Been thinking lately about the ability to do view transitions.
Reviewed my rejected PR (#6893) and tried to come up with a plugin that can achieve this functionality without modifying the `navigate` plugin.

So far, my investigation indicates *it is not possible*, the reason being that any calls to `document.startViewTransition()` only instantiate the [`ViewTransition`](https://developer.mozilla.org/en-US/docs/Web/API/ViewTransition) object and return it, however, the pseudo-element tree is not created immediately with the object, but is fulfilled at a later time via [`ViewTransition.ready`](https://developer.mozilla.org/en-US/docs/Web/API/ViewTransition#instance_properties) property (a `Promise`). Here is a rough diagram of how things currently work together:

```mermaid
sequenceDiagram
  autonumber

  participant al as Alpine
  participant doc as Document
  participant vt as View Transition

  note over al: Navigation triggered

  al->>doc: [Event] alpine:navigate
  activate al
  note over al: ⬇️ Fetch HTML
  al->>doc: [Event] alpine:navigating
  deactivate al
  doc-->>vt: 🔄 Start View Transition
  vt-->>doc: 🫡 View Transition Object
  activate vt
  note over doc,vt: Pseudo-element tree not ready

  activate al
  note over al: 🔄 Prepare for Navigation
  al-->>doc: 🏗️ Morph HTML
  note over al,doc: ✨ Navigation
  deactivate al
  
  al->>doc: [Event] alpine:navigated

  vt-->>doc: 🛫 View Transition Ready
  deactivate vt
  activate doc
  note over doc: 🤷 No changes detected
  vt-->>doc: 🛬 View Transition Finished
  deactivate doc
```

For a view transition plugin to be possible, we would need to have a way to `await` for `ViewTransition.ready` to fulfill before processing the page changes.

The implementation from the PR would still be valid as it does not care about the details of what happens in the `onNavigation()` hooks, and only wraps the logic of updating the HTML in a `Promise.allSettled()` in case there is any promise to handle.
With this ability to `await` before updating the html, it would now be 100% possible to integrate view transitions, either manually or via a plugin:

```mermaid
sequenceDiagram
  autonumber

  participant al as Alpine
  participant doc as Document
  participant vt as View Transition

  note over al: Navigation triggered

  al->>doc: [Event] alpine:navigate
  activate al
  note over al: ⬇️ Fetch HTML
  al->>doc: [Event] alpine:navigating
  deactivate al

  doc-->>vt: 🔄 Start View Transition
  vt-->>doc: 🫡 View Transition Object
  activate vt
  note over doc,vt: Pseudo-element tree not ready
  al-->vt: 📡 Wait for Promises to Resolve

  vt-->al: 🛫 View Transition Ready
  deactivate vt
  activate doc

  activate al
  note over al: 🔄 Prepare for Navigation

  al-->>doc: 🏗️ Morph HTML
  note over al,vt: ✨ Navigation + 🕺 Transition
  deactivate al
  
  al->>doc: [Event] alpine:navigated

  vt-->>doc: 🛬 View Transition Finished
  deactivate doc
```

@calebporzio do any ideas come to mind of a way to register (via hooks or any other way)  and handle `Promise`s?

**Comment by RealMrHex** [comment]
@calebporzio up +1

**Comment by vlados** [comment]
Why this feature is not having any progress? This will make any livewire site more attractive ...
What I can do? Any suggestions???

**Comment by RealMrHex** [comment]
im waiting too... =)
there is no clean way to have it out of the box @vlados 
it need code base modifications and u just need to wait till livewire maintainers decide to merge it.

**Reply by vlados** [reply]
@calebporzio

**Reply by RealMrHex** [reply]
@calebporzio any update?

**Reply by RealMrHex** [reply]
🤷‍♂️

**Comment by BorisKM** [comment]
In my opinion, it's enough just to add `event.preventDefault()` support in `livewire:navigating` event.

Becouse currently I do it like this:

```js
window.Alpine.data('content', (timeout) => ({
    show: false,
    shouldNavigate: false,

    init() {
        document.addEventListener('livewire:navigate', async (event) => {
            if (this.shouldNavigate) {
                return this.shouldNavigate = false
            } else {
                event.preventDefault()
            }

            this.show = false
            this.shouldNavigate = true

            await new Promise(resolve => setTimeout(resolve, timeout / 2))

            // noinspection ES6MissingAwait
            Livewire.navigate(event.detail.url)
        }, { once: true })

        document.addEventListener('livewire:navigated', () => {
            this.show = true
        }, { once: true })
    },
}))
```

And just add x-transition

**Reply by BorisKM** [reply]
@calebporzio

**Reply by vlados** [reply]
@calebporzio why you are ignoring this???

**Comment by alignwebs** [comment]
We need this now!

**Comment by nickwild-999** [comment]
This would be great any news please?

**Comment by NimaKhadivar** [comment]
we need in v4

**Comment by Uhasith** [comment]
+1 for this

**Comment by xdn27** [comment]
We need this fancy feature +1
Are you have plan to make it as a plugin @iksaku?

**Reply by iksaku** [reply]
As [I mentioned previously](https://github.com/livewire/livewire/discussions/6329#discussioncomment-8990282), the current architecture of the `navigate` plugin doesn’t implement a mechanism to handle promises outside of their own navigation callbacks.

I attempted to provide both the mechanisms necessary and direct support for view transitions in #6893, however, there didn’t seem to be any interest at the time for the feature.

I’m open to give it another try if there’s enough interest in it from the maintainers’ side of things :+1:

**Comment by RealMrHex** [comment]
🦥

**Comment by iksaku** [comment]
It looks like a new API is coming in Livewire v4 that supports view transitions 🎉 

[YouTube Link](https://youtu.be/0AvfYxJ5Ss0?si=itemEeU1AQ-nyve_)

Not sure if this works for `wire:navigate` or if it has a JS API to which we can connect, but I believe it is a first step in the right direction!

Thanks @calebporzio 🙏

**Reply by dorgan** [reply]
Based on my current experience it still doesn't support this.
